src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4279,18 +4385,60 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
                 // In the when-null branch, the receiver is known to be maybe-null.
                 // In the other branch, the receiver is known to be non-null.
                 LearnFromNullTest(receiver, ref receiverState);
-                LearnFromNonNullTest(receiver, ref this.State);
-                var nextConditionalAccessSlot = MakeSlot(receiver);
-                if (nextConditionalAccessSlot > 0 && receiver.Type?.IsNullableType() == true)
-                    nextConditionalAccessSlot = GetNullableOfTValueSlot(receiver.Type, nextConditionalAccessSlot, out _);
+                makeAndAdjustReceiverSlot(receiver);
+            }
 
-                _lastConditionalAccessSlot = nextConditionalAccessSlot;
+            // We want to preserve stateWhenNotNull from accesses in the same "chain":
+            // a?.b(out x)?.c(out y); // expected to preserve stateWhenNotNull from both ?.b(out x) and ?.c(out y)
+            // but not accesses in nested expressions:
+            // a?.b(out x, c?.d(out y)); // expected to preserve stateWhenNotNull from a?.b(out x, ...) but not from c?.d(out y)
+            BoundExpression expr = node.AccessExpression;
+            while (expr is BoundConditionalAccess innerCondAccess)
+            {
+                // we assume that non-conditional accesses can never contain conditional accesses from the same "chain".
+                // that is, we never have to dig through non-conditional accesses to find and handle conditional accesses.
+                VisitRvalue(innerCondAccess.Receiver);
+                _currentConditionalReceiverVisitResult = _visitResult;
+                makeAndAdjustReceiverSlot(innerCondAccess.Receiver);
+
+                // The receiverState here represents the scenario where 0 or more of the access expressions could have been evaluated.
```

Comment:
nit: I don't have a great idea, but "receiverState" doesn't seem such a good name now...

Reply:
`stateWhenNull` perhaps :)

Reply:
I think I'll go with `savedState` to match the corresponding place in `AbstractFlowPass`.